doc/developer/design/20230330_recursion_limit.md

```diff
@@ -0,0 +1,159 @@
+
+The syntax follows [our SQL Design Principles](https://www.notion.so/materialize/SQL-design-principles-2e2069797f7c46b8b4c7dd10043afdcd):
+- The options block is parenthesized.
+- There is a space between ITERATION and LIMIT.
+- We accept an optional `=` before the number.
+- We use the `generate_extracted_config!` macro to process the options.
+
+The exact keywords are not final yet, and also depends on what the exact keyword will be for WITH MUTUALLY RECURSIVE.
+
+## Rendering
+
+The rendering part of the design is simple and [uncontroversial](https://materializeinc.slack.com/archives/C041KPFCR98/p1679655472232159?thread_ts=1679653114.252349&cid=C041KPFCR98). We will use `branch_when` ([from Timely](https://github.com/TimelyDataflow/timely-dataflow/blob/master/timely/src/dataflow/operators/branch.rs#L94)) to access the pointstamps containing the iteration numbers (after [the consolidation of the result of an iteration](https://github.com/MaterializeInc/materialize/blob/26f23fd7271cf92c7c8b0dc4ba4711999b45999a/src/compute/src/render/mod.rs#L643)). This will give us two streams: rows that are within the limit, and rows that are outside. The latter one we can either throw away or route into the error stream, depending on whether we want to just gracefully stop at the limit with the current state as the final result (soft limit), or error out when reaching the limit (hard limit).
+
+## IRs
+
+At the AST and HIR levels we can simply add a field to our WMR block representation. However, in MIR we have the problem that if we simply put one limit into each WMR block, then limit settings would not survive `NormalizeLets`. This is because `NormalizeLets` sometimes merges WMR blocks (and moves bindings around in some other ways). For example, if there are two independent WMR clauses (whose results eventually flow together into, e.g., a union or join), then `NormalizeLets` merges these two WMR blocks. In this case, setting just one limit of this merged WMR clause would be wrong.
+
+To solve the above problem, I propose to have a per-let-binding limit in MIR and LIR (and in rendering), so that different let bindings that came from different WMR clauses with differing limits could retain their differing limits. This needs some improvements in `NormalizeLets`:
+- We need to attend to the 1 `LetRec` construction in `NormalizeLets`: at the end of `digest_lets`. We need to come up with limits for each element of `bindings`. What goes into `bindings` is what `digest_lets_helper` puts into `bindings` and `worklist`. Smartening the insertion into `worklist` is easy, because that is a non-recursive `Let`, so the limit can be infinite. Smartening the insertion into `bindings` is also easy, because the binding directly comes from a `LetRec` binding, from which we could copy the limit.
```

Comment:
```suggestion
- We need to attend to the `LetRec` construction in `NormalizeLets`: at the end of `digest_lets`. We need to come up with limits for each element of `bindings`. What goes into `bindings` is what `digest_lets_helper` puts into `bindings` and `worklist`. Smartening the insertion into `worklist` is easy, because that is a non-recursive `Let`, so the limit can be infinite. Smartening the insertion into `bindings` is also easy, because the binding directly comes from a `LetRec` binding, from which we could copy the limit.
```
Stray?

Reply:
changed the text a bit